tsl/src/bgw_policy/job.c

```diff
@@ -513,24 +520,78 @@ policy_invoke_recompress_chunk(Chunk *chunk)
 bool
 policy_compression_execute(int32 job_id, Jsonb *config)
 {
-	int32 chunkid;
+	List *chunkid_lst;
+	ListCell *lc;
 	const Dimension *dim;
 	PolicyCompressionData policy_data;
+	bool distributed, used_portalcxt = false, verbose_log;
+	MemoryContext saved_cxt, multitxn_cxt;
 
 	policy_compression_read_and_validate_config(config, &policy_data);
 	dim = hyperspace_get_open_dimension(policy_data.hypertable->space, 0);
-	chunkid = get_chunk_to_compress(dim, config);
+	distributed = hypertable_is_distributed(policy_data.hypertable);
+	verbose_log = policy_compression_get_verbose_log(config);
+	/* we want the chunk id list to survive across transactions. So alloc in
+	 * a different context
+	 */
+	if (PortalContext)
+	{
+		/*if we have a portal context use that - it will get freed automatically*/
+		multitxn_cxt = PortalContext;
+		used_portalcxt = true;
```

Comment:
I don't see in postgresql source code a place where `PortalContext` is zeroed. Is it possible for postgresql to destroy `PortalContext` before `used_portalcxt = true` is set?

Reply:
We can invoke the registered procedure in 2 ways 1) bgw automatically runs it or 2) CALL run-job(<jobid>).
This handles case 2.
PortalContext is valid for the entire duration of the CALL stmt.